### PR TITLE
[package] fix whichmodule for OrderedImporter

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55646 [package] fix whichmodule for OrderedImporter**

Differential Revision: [D27664238](https://our.internmc.facebook.com/intern/diff/D27664238)

### CODE REVIEW

File: torch/package/importer.py [hunk 1]
```diff
@@ -184,3 +184,11 @@ def import_module(self, module_name: str) -> ModuleType:
             raise last_err
         else:
             raise ModuleNotFoundError(module_name)
+
+    def whichmodule(self, obj: Any, name: str) -> str:
+        for importer in self._importers:
+            module_name = importer.whichmodule(obj, name)
```
Comment (SplitInfinity) [inline torch/package/importer.py:190]: Might the importer implementation raise `ModuleNotFoundError`?

File: torch/package/importer.py [hunk 2]
```diff
@@ -184,3 +184,11 @@ def import_module(self, module_name: str) -> ModuleType:
...
+            if module_name != "__main__":
+                return module_name
+
+        return "__main__"
```
Comment (SplitInfinity) [inline torch/package/importer.py:194]: This is what the CPython `whichmodule` does too but I don't really understand why. 😕  Is it understood to indicate that a module wasn't found?

File: test/package/test_importer.py [hunk 3]
```diff
@@ -84,6 +85,32 @@ def test_ordered_importer_basic(self):
...
+        class DummyClass:
+            pass
+
+        dummy_importer_foo = DummyImporter("foo")
+        dummy_importer_bar = DummyImporter("bar")
+        foo_then_bar = OrderedImporter(dummy_importer_foo, dummy_importer_bar)
+        self.assertEqual(foo_then_bar.whichmodule(DummyClass(), ""), "foo")
+
+        bar_then_foo = OrderedImporter(dummy_importer_bar, dummy_importer_foo)
+        self.assertEqual(bar_then_foo.whichmodule(DummyClass(), ""), "bar")
```
Comment (SplitInfinity) [inline test/package/test_importer.py:112]: Consider adding a case in which the first importer fails to find the module that the object belongs to.

**Review by SplitInfinity — APPROVED** [review]

File: torch/package/importer.py (hunk 2 above)
Comment (suo) [reply torch/package/importer.py:194]: honestly I am not sure, I suppose the idea is that an object without `__module__` defined that also isn't accessible from any known module *must* have been defined interactively/from a script.

File: torch/package/importer.py (hunk 1 above)
Comment (suo) [reply torch/package/importer.py:190]: `whichmodule` is not supposed to actually import any modules or have side effects into the import system, so I don't think so.

**Pull request closed without merging**